File: src/main/java/frc/robot/subsystems/CANDriveSubsystem.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.subsystems;

import org.fairportrobotics.frc.posty.TestableSubsystem;
import org.fairportrobotics.frc.posty.test.PostTest;
import static org.fairportrobotics.frc.posty.assertions.Assertions.*;

import com.ctre.phoenix.motorcontrol.can.WPI_VictorSPX;
import com.ctre.phoenix.WPI_CallbackHelper;

import com.revrobotics.spark.SparkBase.PersistMode;
import com.revrobotics.spark.SparkBase.ResetMode;
import com.revrobotics.spark.SparkLowLevel.MotorType;
import com.revrobotics.spark.SparkMax;
import com.revrobotics.spark.config.SparkMaxConfig;

import edu.wpi.first.wpilibj.drive.DifferentialDrive;
import edu.wpi.first.wpilibj.motorcontrol.Victor;
import frc.robot.Constants.DriveConstants;

// Class to drive the robot over CAN
public class CANDriveSubsystem extends TestableSubsystem {
  private final WPI_VictorSPX leftLeader;
  private final WPI_VictorSPX leftFollower;
  private final WPI_VictorSPX rightLeader;
  private final WPI_VictorSPX rightFollower;

  private final DifferentialDrive drive;

  public CANDriveSubsystem() {
    super();
    // create brushed motors for drive
    leftLeader = new WPI_VictorSPX(DriveConstants.LEFT_LEADER_ID);
    leftFollower = new WPI_VictorSPX(DriveConstants.LEFT_FOLLOWER_ID);
    rightLeader = new WPI_VictorSPX(DriveConstants.RIGHT_LEADER_ID);
    rightFollower = new WPI_VictorSPX(DriveConstants.RIGHT_FOLLOWER_ID);
    leftLeader.setInverted(true);
    leftFollower.setInverted(true);

    // set up differential drive class
    drive = new DifferentialDrive(leftLeader, rightLeader);

    // Set can timeout. Because this project only sets parameters once on
    // construction, the timeout can be long without blocking robot operation. Code
    // which sets or gets parameters during operation may need a shorter timeout.
    //leftLeader.setCANTimeout(250);
    //rightLeader.setCANTimeout(250);
    //leftFollower.setCANTimeout(250);
    //rightFollower.setCANTimeout(250);

    // Create the configuration to apply to motors. Voltage compensation
    // helps the robot perform more similarly on different
    // battery voltages (at the cost of a little bit of top speed on a fully charged
    // battery). The current limit helps prevent tripping
    // breakers.
    SparkMaxConfig config = new SparkMaxConfig();
    config.voltageCompensation(12);
    config.smartCurrentLimit(DriveConstants.DRIVE_MOTOR_CURRENT_LIMIT);

    leftFollower.follow(this.leftLeader);
    rightFollower.follow(this.rightLeader);
  }

  @Override
  public void periodic() {
  }

  // sets the speed of the drive motors
  public void driveArcade(double xSpeed, double zRotation) {
    drive.arcadeDrive(xSpeed, zRotation);
  }

  @PostTest
  public void canDevicesConnected(){
    assertThat(rightLeader.getFirmwareVersion()).isGreaterThan(5);
    assertThat(rightFollower.getFirmwareVersion()).isGreaterThan(5);
    assertThat(leftLeader.getFirmwareVersion()).isGreaterThan(5);
    assertThat(leftFollower.getFirmwareVersion()).isGreaterThan(5);
  }

}

File: subsystems/CANDriveSubsystem.py
# Copyright (c) FIRST and other WPILib contributors.
# Open Source Software you can modify and/or share it under the terms of
# the WPILib BSD license file in the root directory of this project.

import commands2
import constants
import rev
import wpilib
import wpilib.drive
from magicbot import feedback


class CANDriveSubsystem(commands2.Subsystem):
    # The next three lines were added for the genie bot
    MAX_OUTPUT = 0.0
    ROTATION = 0.0
    SPEED = 0.0

    def __init__(self) -> None:
        super().__init__()

        # spark max motor controllers in brushed mode
        self.leftLeader = rev.SparkMax(
            constants.LEFT_LEADER_ID, rev.SparkBase.MotorType.kBrushed
        )
        self.leftFollower = rev.SparkMax(
            constants.LEFT_FOLLOWER_ID, rev.SparkBase.MotorType.kBrushed
        )
        self.rightLeader = rev.SparkMax(
            constants.RIGHT_LEADER_ID, rev.SparkBase.MotorType.kBrushed
        )
        self.rightFollower = rev.SparkMax(
            constants.RIGHT_FOLLOWER_ID, rev.SparkBase.MotorType.kBrushed
        )

        # this is the differential drive instance which allows us to control
        # the drive with joysticks
        self.drive = wpilib.drive.DifferentialDrive(self.leftLeader, self.rightLeader)

        # set can timeouts. This program only sets parameters on startup and
        # doesn't get any parameters so a long timeout is acceptable. Programs
        # which set or get parameters during runtime likely want a timeout
        # closer or equal to the default.
        self.leftLeader.setCANTimeout(constants.CAN_TIMEOUT)
        self.rightLeader.setCANTimeout(constants.CAN_TIMEOUT)
        self.leftFollower.setCANTimeout(constants.CAN_TIMEOUT)
        self.rightFollower.setCANTimeout(constants.CAN_TIMEOUT)

        self.sparkConfig = rev.SparkMaxConfig()

        # enable voltage compensation. This makes the performance more consistent
        # at different levels of battery charge at the cost of some peak performance
        # with a fully charged battery
        self.sparkConfig.voltageCompensation(constants.VOLTAGE_COMPENSATION)

        # set current limit. This helps prevent tripping breakers
        self.sparkConfig.smartCurrentLimit(constants.DRIVE_MOTOR_CURRENT_LIMIT)

        # set to follow leader and then use to configure corresponding follower
        self.sparkConfig.follow(self.leftLeader)
        self.leftFollower.configure(
            self.sparkConfig,
            rev.SparkBase.ResetMode.kResetSafeParameters,
            rev.SparkBase.PersistMode.kPersistParameters,
        )
        self.sparkConfig.follow(self.rightLeader)
        self.rightFollower.configure(
            self.sparkConfig,
            rev.SparkBase.ResetMode.kResetSafeParameters,
            rev.SparkBase.PersistMode.kPersistParameters,
        )

        # disable following and use to configure leader. Invert before configuring
        # left side so that postive values drive both sides forward
        self.sparkConfig.disableFollowerMode()
        self.rightLeader.configure(
            self.sparkConfig,
            rev.SparkBase.ResetMode.kResetSafeParameters,
            rev.SparkBase.PersistMode.kPersistParameters,
        )
        self.sparkConfig.inverted(True)
        self.leftLeader.configure(
            self.sparkConfig,
            rev.SparkBase.ResetMode.kResetSafeParameters,
            rev.SparkBase.PersistMode.kPersistParameters,
        )

    # function to drive with joystick inputs
    def arcadeDrive(self, xSpeed: float, zRotation: float) -> None:
        # self.drive.arcadeDrive(xSpeed, zRotation)

        # =====================================================================
        # ALL LINES BELOW THIS COMMENT WERE ADDED FOR THE GENIE BOT
        # =====================================================================

        self.SPEED = self.MAX_OUTPUT * xSpeed
        self.ROTATION = self.MAX_OUTPUT * zRotation
        self.drive.arcadeDrive(self.SPEED, self.ROTATION, squareInputs=True)

    def execute(self):
        pass

    def set_max_output(self, max_output: float) -> None:
        """
        Set the maximum output of the drive.

        :param max_output: The maximum output to set, between 0.0 and 1.0.
        """
        if not 0.0 <= max_output <= 1.0:
            raise ValueError("Max output must be between 0.0 and 1.0")
        self.MAX_OUTPUT = max_output

    def get_drive_type(self) -> str:
        """Get the current drive type."""
        return "arcade"

    def stop(self) -> None:
        """Stop the drivetrain."""
        self.drive.stopMotor()

    @feedback(key="Max Output")
    def get_max_output(self) -> float:
        """Get the maximum output of the drive."""
        return self.MAX_OUTPUT

    @feedback(key="Speed")
    def get_speed(self) -> float:
        """Get the speed passed into the drive."""
        return self.SPEED

    @feedback(key="Rotation")
    def get_rotation(self) -> float:
        """Get the rotation passed into the drive."""
        return self.ROTATION
